packages/babel-plugin-proposal-import-defer/src/index.ts

```diff
@@ -0,0 +1,121 @@
+import { declare } from "@babel/helper-plugin-utils";
+import type { types as t } from "@babel/core";
+import type { Scope } from "@babel/traverse";
+import { defineCommonJSHook } from "@babel/plugin-transform-modules-commonjs";
+
+import syntaxImportDefer from "@babel/plugin-syntax-import-defer";
+
+export default declare(api => {
+  api.assertVersion(7);
+  // We need the explicit type annotation otherwise when using t.assert* ts
+  // reports that 'Assertions require every name in the call target to be
+  // declared with an explicit type annotation'
+  const t: typeof api.types = api.types;
+  const { template } = api;
+
+  function allReferencesAreProps(scope: Scope, node: t.ImportDeclaration) {
+    const specifier = node.specifiers[0];
+    t.assertImportNamespaceSpecifier(specifier);
+
+    const binding = scope.getOwnBinding(specifier.local.name);
+    return !!binding?.referencePaths.every(path =>
+      path.parentPath.isMemberExpression({ object: path.node }),
+    );
+  }
+
+  return {
+    name: "proposal-import-defer",
+
+    inherits: syntaxImportDefer,
+
+    pre() {
+      const { file } = this;
+
+      defineCommonJSHook(file, {
+        name: PACKAGE_JSON.name,
+        version: PACKAGE_JSON.version,
+        getWrapperPayload(source, metadata, importNodes) {
+          let needsProxy = false;
+          for (const node of importNodes) {
+            if (!t.isImportDeclaration(node)) return null;
+            if (node.phase !== "defer") return null;
+            if (!allReferencesAreProps(file.scope, node)) needsProxy = true;
+          }
+          return needsProxy ? "defer/proxy" : "defer/function";
+        },
+        buildRequireWrapper(name, init, payload, referenced) {
+          if (payload === "defer/proxy") {
+            if (!referenced) return false;
+            return template.statement.ast`
+              var ${name} = ${file.addHelper("importDeferProxy")}(
+                () => ${init}
+              )
+            `;
+          }
+          if (payload === "defer/function") {
+            if (!referenced) return false;
+            return template.statement.ast`
+              function ${name}() {
+                const data = ${init};
```

Comment:
The memoizer variable can also be injected from the function arguments.

Reply:
Yeah we could do `function ${name}(data = ${init})`, but then compiling that down will generate bigger code because it has to check if `data` is undefined.

Reply:
```
function ${name}(data) {
  ${name} = () => data;
  return data = ${init};
}
```
:)